src/pip/_internal/operations/prepare.py

```diff
@@ -343,7 +358,9 @@ def _get_linked_req_hashes(self, req: InstallRequirement) -> Hashes:
         # report less-useful error messages for unhashable
         # requirements, complaining that there's no hash provided.
         if req.link.is_vcs:
-            raise VcsHashUnsupported()
+            if not req.user_supplied:
+                raise DependencyVcsHashNotSupported()
+            return VcsHashes()
         if req.link.is_existing_dir():
             raise DirectoryUrlHashUnsupported()
```

Comment:
Note to self: when `req.is_wheel_from_cache` is True, link is the cached wheel and not the link to the artifact that was downloaded to build the wheel. So this function may not necessarily do what one expects when reasoning on req.link.